packages/next/src/middleware/sitemaps/createSitemaps.ts

```diff
@@ -51,21 +52,54 @@ const parserConfig: Partial<X2jOptions> = {
  * @returns {Response|undefined}
  */
 export async function createRootSitemapIndex(
-  req: NextRequest,
+  req: NextRequest | IncomingMessage,
   normalizedConfig: NormalizedConfig,
+  isMiddleware = true,
 ): Promise<Response | undefined> {
   const { pages, sitemapPathsToIgnore, replaceUrls, wpUrl } = normalizedConfig;
-  const { pathname, origin } = new URL(req.url);
+
+  if (!req.url) {
+    throw new Error('Request object must have URL');
+  }
+
+  let wpSitemapUrl = '';
+  let frontendUrl = '';
+  if (isMiddleware) {
+    const { pathname, origin } = new URL(req.url);
+    frontendUrl = origin;
+    wpSitemapUrl = `${trimSlashes(wpUrl)}/${trimSlashes(pathname)}`;
+  } else {
+    // get rootSitemapPath config param
+    // fetch sitemap from WP
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+    wpSitemapUrl = `${trimSlashes(wpUrl)}/${trimSlashes(
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+      normalizedConfig.rootSitemapPath,
+    )}`;
+    frontendUrl = normalizedConfig.frontendUrl;
+  }
+
   let sitemaps: SitemapSchemaSitemapElement[] = [];
 
   if (!isUndefined(pages) && isArray(pages) && pages.length) {
+    let sitemapFaustPagesUrl = '';
+    if (isMiddleware) {
+      sitemapFaustPagesUrl = `${trimSlashes(frontendUrl)}/${trimSlashes(
+        FAUST_PAGES_PATHNAME,
+      )}`;
+    } else {
+      sitemapFaustPagesUrl = `${trimSlashes(
+        frontendUrl,
+      )}/sitemap.xml?sitemap=${trimSlashes(FAUST_PAGES_PATHNAME)}`;
+    }
     sitemaps = [
       ...sitemaps,
-      { loc: `${trimSlashes(origin)}/${trimSlashes(FAUST_PAGES_PATHNAME)}` },
+      {
+        loc: sitemapFaustPagesUrl,
```

Comment:
We should run this through `encodeURI()`:

https://www.sitemaps.org/protocol.html#locdef